README.md
# Nova Rating field

A **Star rating field** to use in your Laravel Nova apps. 

Uses [vue-star-rating](https://github.com/craigh411/vue-star-rating). 

[![Latest Version on Packagist](https://img.shields.io/packagist/v/nikaia/nova-rating-field.svg?style=flat-square)](https://packagist.org/packages/nikaia/nova-rating-field)
[![Total Downloads](https://img.shields.io/packagist/dt/nikaia/nova-rating-field.svg?style=flat-square)](https://packagist.org/packages/nikaia/nova-rating-field)


<p align="center">
    <img src="./docs/players-index.png"/>
    <br/>
    <img src="./docs/players-edit.png" />
</p>


## Installation

You can install the package in to a Laravel app that uses [Nova](https://nova.laravel.com) via composer:

```bash
composer require nikaia/nova-rating-field
```

Next you can use the `Nikaia\Rating\Rating` field in your Nova resource.

## Usage

```php
public function fields(Request $request)
{
    return [
        // ...

        Rating::make('Rating')->min(0)->max(5)->increment(0.5)->hideFromIndex(),
        
        // Defining a custom style for the index page.
        Rating::make('Rating')->min(0)->max(5)->increment(0.5)->hideRating()
            ->withStyles([
                'star-size' => 15,
                'rounded-corners' => true,
            ])->onlyOnIndex()->sortable(),

        // ...    
        
    ];
}
```

### Defining properties

```php
public function fields(Request $request)
{
    Rating::make('Rating')
    
        // Miniumum rating (default: 0)
        ->min(0) 
        
        // Maximum rating (default: 5)
        // This is how the component knows how many stars it should display.
        ->max(5)
        
        // Incremet (default: 1)
        // Can be float. The underlying eloquent colum must be defined as float in that case.
        // ie. 0.5 for half stars or 0.01 for fluid stars.
        ->increment(0.5)
        
        // Show rating value next to the stars
        ->hideRating()
        
}
```

### Customizing styles

You can style the component using `withStyles` method. Options are passed to the the underlying [vue component style props](https://github.com/craigh411/vue-star-rating#style-props). 

Default values are :

```php
public function fields(Request $request)
{
    Rating::make('Rating')
        ->withStyles([
            'star-size' => 30,
            'active-color' => 'var(--primary)', // Primary nova theme color.
            'inactive-color' => '#d8d8d8',
            'border-color' => 'var(--60)',
            'border-width' => 0,
            'padding' => 10,
            'rounded-corners' => false,
            'inline' => false,
            'glow' => 0,
            'glow-color' => '#fff',
            'text-class' => 'inline-block text-80 h-9 pt-2',
        ]);
}
```


CHANGELOG.md
# Changelog

All notable changes to `nova-rating-field` will be documented in this file

## 1.0.0 - 2018-09-15

- initial release


tests/TestCase.php
<?php

namespace Nikaia\Rating\Tests;

use Illuminate\Support\Facades\Route;
use Orchestra\Testbench\TestCase as Orchestra;
use Nikaia\Rating\ToolServiceProvider;

abstract class TestCase extends Orchestra
{
    public function setUp()
    {
        parent::setUp();

        Route::middlewareGroup('nova', []);
    }

    protected function getPackageProviders($app)
    {
        return [
            ToolServiceProvider::class,
        ];
    }
}


tests/ToolControllerTest.php
<?php

namespace Nikaia\Rating\Tests;

use Nikaia\Rating\Http\Controllers\ToolController;
use Nikaia\Rating\Rating;
use Symfony\Component\HttpFoundation\Response;

class ToolControllerTest extends TestCase
{
    /** @test */
    public function it_can_can_return_a_response()
    {
        $this
            ->get('nova-vendor/nikaia/nova-rating-field/endpoint')
            ->assertSuccessful();
    }
}


src/Rating.php
<?php

namespace Nikaia\Rating;

use Laravel\Nova\Fields\Field;

class Rating extends Field
{
    /**
     * The field's component.
     *
     * @var string
     */
    public $component = 'rating';

    /**
     * Default component styles.
     *
     * @var array
     */
    protected static $defaultStyles = [
        'star-size' => 30,
        'active-color' => 'var(--primary)',
        'inactive-color' => '#d8d8d8',
        'border-color' => 'var(--60)',
        'border-width' => 0,
        'padding' => 10,
        'rounded-corners' => false,
        'inline' => false,
        'glow' => 0,
        'glow-color' => '#fff',
        'text-class' => 'inline-block text-80 h-9 pt-2',
    ];

    /**
     * Rating constructor.
     *
     * @param string $name
     * @param null|string $attribute
     * @param mixed|null $resolveCallback
     */
    public function __construct(string $name, ?string $attribute = null, ?mixed $resolveCallback = null)
    {
        parent::__construct($name, $attribute, $resolveCallback);

        $this->withMeta(['min' => 0, 'max' => 5]);
        $this->withStyles(static::$defaultStyles);
    }

    /**
     * Minimum rating.
     * default = 0.
     *
     * @param $min
     * @return Rating
     */
    public function min($min)
    {
        return $this->withMeta(['min' => $min]);
    }

    /**
     * @param $max int Maximum rating.
     * This will let the component know how many stars it should display.
     *
     * @return Rating
     */
    public function max($max)
    {
        return $this->withMeta(['max' => $max]);
    }

    /**
     * @param float The rating increment
     * for example pass 0.5 for half stars or 0.01 for fluid stars.
     * Expects a number between 0.01 - 1.
     *
     * @return Rating
     */
    public function increment($increment)
    {
        return $this->withMeta(['increment' => $increment]);
    }

    /**
     * Show rating value next to stars.
     * this is the default behaviour.
     *
     * @return Rating
     */
    public function showRating()
    {
        return $this->withMeta(['showRating' => true]);
    }

    /**
     * Hide rating value next to stars.
     * By default rating will be displayed unless you use this method.
     *
     * @return Rating
     */
    public function hideRating()
    {
        return $this->withMeta(['showRating' => false]);
    }

    /**
     * Style the component.
     *
     * @param array $styles
     * @return Rating
     */
    public function withStyles(array $styles)
    {
        $build = [];
        foreach (static::$defaultStyles as $key => $defaultValue) {
            $build[$key] = array_get($styles, $key, $defaultValue);
        }

        return $this->withMeta($build);
    }
}
